File: 프로그래머스/3/12942. 최적의 행렬 곱셈/최적의 행렬 곱셈.java
import java.util.Arrays;

class Solution {
    
    private static final int INFINITY = 1_000_000_000;
    
    public int solution(int[][] matrixSizes) {
        int n = matrixSizes.length;
        int[][] dp = new int[n][n];
        
        for (int i = 0; i < n; i++) {
            Arrays.fill(dp[i], INFINITY);
            dp[i][i] = 0;
        }
        
        for (int length = 1; length < n; length++) {
            for (int i = 0; i < n - length; i++) {
                int j = i + length;
                
                for (int k = i; k < j; k++) {
                    int cost = dp[i][k] + dp[k + 1][j] + (matrixSizes[i][0] * matrixSizes[k][1] * matrixSizes[j][1]);
                    if (cost < dp[i][j]) {
                        dp[i][j] = cost;
                    }
                }
            }
        }
        
        return dp[0][n - 1];
    }
}

File: 프로그래머스/3/12942. 최적의 행렬 곱셈/최적의 행렬 곱셈.py
def solution(matrix_sizes):
    n = len(matrix_sizes)
    INFINITY = 1_000_000_000

    dp = [[INFINITY] * n for _ in range(n)]
    for i in range(n):
        dp[i][i] = 0

    for length in range(1, n):
        for i in range(n - length):
            j = i + length
            for k in range(i, j):
                cost = (dp[i][k] + dp[k + 1][j] + matrix_sizes[i][0] * matrix_sizes[k][1] * matrix_sizes[j][1])
                if cost < dp[i][j]:
                    dp[i][j] = cost

    return dp[0][n - 1]